### PR TITLE
Fix test Bearer -> Token + refactoring

### PR DESCRIPTION
**Pull request by guillaumemaka**
@JackyChiu I recreate a branch from the upstream/master an re-apply all the changes I made in the articles_features. So hope that fix the issue.

All Test passes 
```
== RUN   TestArticlesHandler_Index
--- PASS: TestArticlesHandler_Index (0.00s)
=== RUN   TestArticlesHandler_Read
--- PASS: TestArticlesHandler_Read (0.00s)
=== RUN   TestArticlesHandler_FilterByTag
--- PASS: TestArticlesHandler_FilterByTag (0.00s)
=== RUN   TestArticlesHandler_FilterByAuthor
--- PASS: TestArticlesHandler_FilterByAuthor (0.00s)
=== RUN   TestArticlesHandler_FilterByFavorited
--- PASS: TestArticlesHandler_FilterByFavorited (0.00s)
=== RUN   TestArticlesHandler_CreateUnauthorized
--- PASS: TestArticlesHandler_CreateUnauthorized (0.00s)
=== RUN   TestArticlesHandler_Create
--- PASS: TestArticlesHandler_Create (0.00s)
=== RUN   TestArticlesHandler_CreateWithEmptyTitle
--- PASS: TestArticlesHandler_CreateWithEmptyTitle (0.00s)
=== RUN   TestArticlesHandler_CreateWithEmptyDescription
--- PASS: TestArticlesHandler_CreateWithEmptyDescription (0.00s)
=== RUN   TestArticlesHandler_CreateWithEmptyBody
--- PASS: TestArticlesHandler_CreateWithEmptyBody (0.00s)
=== RUN   TestArticlesHandler_UpdateForbidden
--- PASS: TestArticlesHandler_UpdateForbidden (0.00s)
=== RUN   TestArticlesHandler_UpdateNotAuthorized
--- PASS: TestArticlesHandler_UpdateNotAuthorized (0.00s)
=== RUN   TestArticlesHandler_UpdateOK
--- PASS: TestArticlesHandler_UpdateOK (0.00s)
=== RUN   TestArticlesHandler_Favorite
--- PASS: TestArticlesHandler_Favorite (0.00s)
=== RUN   TestArticlesHandler_FavoriteAlreadyFavoritedArticle
--- PASS: TestArticlesHandler_FavoriteAlreadyFavoritedArticle (0.00s)
=== RUN   TestArticlesHandler_Unfavorite
--- PASS: TestArticlesHandler_Unfavorite (0.00s)
=== RUN   TestArticlesHandler_UnfavoriteNotYetFavoritedArticle
--- PASS: TestArticlesHandler_UnfavoriteNotYetFavoritedArticle (0.00s)
=== RUN   TestArticlesHandler_DeleteOk
--- PASS: TestArticlesHandler_DeleteOk (0.00s)
=== RUN   TestArticlesHandler_DeleteForbidden
--- PASS: TestArticlesHandler_DeleteForbidden (0.00s)
=== RUN   TestArticlesHandler_DeleteUnauthorized
--- PASS: TestArticlesHandler_DeleteUnauthorized (0.00s)
=== RUN   TestArticlesHandler_ValidTokenButUserNotExist
--- PASS: TestArticlesHandler_ValidTokenButUserNotExist (0.00s)
PASS
ok  	github.com/JackyChiu/realworld-starter-kit/handlers	0.630s
```

### CODE REVIEW

File: models/article.go [hunk 1]
```diff
@@ -184,21 +218,208 @@ func (db *DB) UnfavoriteArticle(userID int, articleID int) error {
...
 	return db.Order("articles.created_at desc").
 		Preload("Tags").
 		Preload("User")
 }
+
+///////////////////////////////////////////////////////////////////////////////
+// Private Methods															 //
+///////////////////////////////////////////////////////////////////////////////
+
+func buildWhereClause(field string, vars ...interface{}) (string, interface{}, bool) {
```
Comment (JackyChiu) [inline models/article.go:372]: What is this function for?

**Review by JackyChiu — COMMENTED** [review]
Thanks for fixing the commit issues and tests! Got a question about a method you added

File: models/article.go (hunk 1 above)
Comment (guillaumemaka) [reply models/article.go:372]: Ex: with the FilterByTag()
I can invoke 
```go
// query is an instance of gorm.DB
FilterByTag(query, []string{"tag1", "tag2"})
// will call buildWhereClause("tags.name", string["tag1", "tag2"])
// whereClause, value, skip
// skip:  
//   true: don't apply the filter (ex: empty string, empty slice,...)
//   false: apply
-> tags.name IN (?), []string{"tag1", "tag2"}, false

FilterByTag(query, "tag1")
// will call buildWhereClause("tags.name", "tag1")
// whereClause, value, skip
-> tags.name = ?, "tag1", false

FilterByTag(query, url.Values{"tag": "tag1")
// will call buildWhereClause("tags.name", "tag", "tag1")
// whereClause, value, skip
-> tags.name = ?, "tag1", false
```
Example in handlers/article.go
```go
r.ParseForm()
query = h.DB.Limit(query, r.Form) -> will lookup the 'limit' key in the map if exists
query = h.DB.Offset(query, r.Form) -> will lookup the 'offset' key in the map if exists
query = h.DB.FilterByTag(query, r.Form) -> will lookup the 'tag' key in the map if exists
query = h.DB.FilterAuthoredBy(query, r.Form) -> will lookup the 'author' key in the map if exists
query = h.DB.FilterFavoritedBy(query, r.Form) -> will lookup the 'favorited' key in the map if exists

err = query.Find(&articles).Error
```


so buildWhereClause() its just an helper function to construct a where clause for a kind of value type:
```go
// primitive value type int, string, slice
buildWhereClause(column, value)

// url.Values (map)
buildWhereClause(column, mapKey, map)
```
but I may change the signature of the different filter later to be more explicit like
```go
h.DB.FilterByTag(query,"tag1") -> query  tags.name = tag1
h.DB.FilterByTag(query, []int{1, 2, 3} -> query the tags.id IN (1,2,3)

// For a map (or url.Values) specify the key to lookup
h.DB.FilterByTag(query,"tag", request.Form) -> use the value of the 'tag' key in the map if exists

// or

// query, column, value
h.DB.FilterByTag(query,"tags.id", []int{1,2,3})
// query, column, mapKey, map
h.DB.FilterByTag(query,"tags.name", "tag", request.Form)
```

Comment (JackyChiu) [reply models/article.go:372]: Alright sounds good

**Review by JackyChiu — APPROVED** [review]

**Pull request merged**